rust/main/agents/relayer/src/msg/pending_message.rs

```diff
@@ -250,99 +252,64 @@ impl PendingOperation for PendingMessage {
+        // first attempt at gas estimation
+        let mut tx_cost_estimate = self
+            .ctx
+            .destination_mailbox
+            .process_estimate_costs(&self.message, &metadata_bytes)
+            .await;
 
-        let metadata_bytes = match message_metadata_builder
-            .build(ism_address, &self.message, params)
-            .await
-        {
-            Ok(metadata) => {
-                let metadata_bytes = metadata.to_vec();
-                self.metadata = Some(metadata_bytes.clone());
-                metadata_bytes
-            }
-            Err(err) => {
-                match &err {
-                    MetadataBuildError::FailedToBuild(_) => {
-                        return self
-                            .on_reprepare(Some(err), ReprepareReason::ErrorBuildingMetadata);
-                    }
-                    MetadataBuildError::CouldNotFetch => {
-                        return self
-                            .on_reprepare::<String>(None, ReprepareReason::CouldNotFetchMetadata);
-                    }
-                    // If the metadata building is refused, we still allow it to be retried later.
-                    MetadataBuildError::Refused(reason) => {
-                        warn!(?reason, "Metadata building refused");
-                        return self
-                            .on_reprepare::<String>(None, ReprepareReason::MessageMetadataRefused);
-                    }
-                    // These errors cannot be recovered from, so we drop them
-                    MetadataBuildError::UnsupportedModuleType(reason) => {
-                        warn!(?reason, "Unsupported module type");
-                        return self
-                            .on_reprepare(Some(err), ReprepareReason::ErrorBuildingMetadata);
-                    }
-                    MetadataBuildError::MaxIsmDepthExceeded(depth) => {
-                        warn!(depth, "Max ISM depth reached");
-                        return self
-                            .on_reprepare(Some(err), ReprepareReason::ErrorBuildingMetadata);
-                    }
-                    MetadataBuildError::MaxIsmCountReached(count) => {
-                        warn!(count, "Max ISM count reached");
-                        return self
-                            .on_reprepare(Some(err), ReprepareReason::ErrorBuildingMetadata);
-                    }
-                    MetadataBuildError::AggregationThresholdNotMet(threshold) => {
-                        warn!(threshold, "Aggregation threshold not met");
-                        return self
-                            .on_reprepare(Some(err), ReprepareReason::CouldNotFetchMetadata);
-                    }
+        // if gas estimate fails, we want to retry building metadata,
+        // and try gas estimation again
+        if tx_cost_estimate.is_err() {
+            self.clear_metadata();
+            match self.build_metadata().await {
```

Comment:
can we make it so we only build metadata twice if the metadata wasn't built this time? i.e. only if it was previously cached